gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java

```diff
@@ -0,0 +1,132 @@
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+/**
+ * {@link DataWriter} that partitions data using a partitioner, instantiates appropriate writers, and sends records to
+ * the chosen writer.
+ */
+public class PartitionedDataWriter<S, D> implements DataWriter<D> {
+
+  private static final GenericRecord NON_PARTITIONED_WRITER_KEY =
+      new GenericData.Record(SchemaBuilder.record("Dummy").fields().endRecord());
+
+  private final Optional<WriterPartitioner> partitioner;
+  private final LoadingCache<GenericRecord, DataWriter<D>> partitionWriters;
+  private final Optional<PartitionAwareDataWriterBuilder> builder;
+  private final boolean shouldPartition;
+
+  public PartitionedDataWriter(DataWriterBuilder<S, D> builder, final State state) throws IOException {
+
+    this.partitionWriters = CacheBuilder.newBuilder().build(new CacheLoader<GenericRecord, DataWriter<D>>() {
+      @Override public DataWriter<D> load(final GenericRecord key) throws Exception {
+        return new InstrumentedPartitionedDataWriterDecorator<D>(createPartitionWriter(key), state, key);
+      }
+    });
+
+    if(state.contains(ConfigurationKeys.WRITER_PARTITIONER_CLASS)) {
+      Preconditions.checkArgument(builder instanceof PartitionAwareDataWriterBuilder,
+          String.format("%s was specified but the writer %s does not support partitioning.",
+              ConfigurationKeys.WRITER_PARTITIONER_CLASS, builder.getClass().getCanonicalName()));
+
+      try {
+        this.shouldPartition = true;
+        this.builder = Optional.of(PartitionAwareDataWriterBuilder.class.cast(builder));
+        this.partitioner = Optional.of(
+            WriterPartitioner.class.cast(ConstructorUtils.invokeConstructor(
+                    Class.forName(state.getProp(ConfigurationKeys.WRITER_PARTITIONER_CLASS)), state)));
+        this.builder.get().validatePartitionSchema(this.partitioner.get().partitionSchema());
+      } catch(ReflectiveOperationException roe) {
+        throw new IOException(roe);
+      }
+    } else {
+      this.shouldPartition = false;
+      this.partitionWriters.put(NON_PARTITIONED_WRITER_KEY,
+          new InstrumentedDataWriterDecorator<D>(builder.build(), state));
+      this.partitioner = Optional.absent();
+      this.builder = Optional.absent();
+    }
+  }
+
+  @Override public void write(D record) throws IOException {
+    try {
+      GenericRecord partition = this.shouldPartition ? this.partitioner.get().partitionForRecord(record) :
+          NON_PARTITIONED_WRITER_KEY;
+      DataWriter<D> writer = this.partitionWriters.get(partition);
+      writer.write(record);
+    } catch(ExecutionException ee) {
+      throw new IOException(ee);
+    }
+  }
+
+  @Override public void commit() throws IOException {
+    for(Map.Entry<GenericRecord, DataWriter<D>> entry : this.partitionWriters.asMap().entrySet()) {
+      entry.getValue().commit();
```

Comment:
I suggest putting `commit` into a `try-catch` block so the commit failure of one partition will not cause other partitions to not be committed.